### PR TITLE
[labs/ssr] fix ModuleLoader failing with multiple concurrent entrypoints

### PR DESCRIPTION
**Pull request by AndrewJakubowicz**
Fixes: https://github.com/lit/lit/issues/2549

### Context

When the eleventy ssr plugin loads modules concurrently, it would previously fail with an error. Something like:

```
Error: request for '@lit/reactive-element' is not yet fulfilled
        at SourceTextModule.link (node:internal/vm/module:198:17)
        at ModuleLoader.importModule (/home/aomarks/work/wireit-lit/packages/labs/ssr/src/lib/module-loader.ts:107:7)
        at async Promise.all (index 1)
```

Investigating this issue I found:
 - Specification of the module methods: https://tc39.es/ecma262/#table-abstract-methods-of-module-records
 - Located the error message in the node codebase: https://github.com/nodejs/node/blob/eed33c9deaa336e76c5385a077490107c6110f08/src/module_wrap.cc#L532-L535

Solution is to closer match the spec. `LoadRequestedModules` should prepare module for linking by loading **all** dependencies.
Then when `Link` is called, `LoadRequestedModules` must have completed.

### Theory

Before linking our code does:

```
async importModule(...){
    ...
    const result = await this._loadModule(specifier, referrerPath); // <--- `LoadRequestedModules`
    const module = result.module as vm.Module;
    if (module.status === 'unlinked') {
      await module.link(this._linker); // <--- `Link`
    }
   ...
}
```

However previously our `_loadModule` method did not load all dependencies, relying on `module.link` to load the dependencies. This seems to work when not loading multiple modules concurrently.

But, to allow multiple modules to load concurrently I've added some additional module loading logic to `_loadModule` so dependencies get loaded.

The result is that by the time `module.link` is reached, all modules should have been loaded.

### Test plan

I added a unit test which throws `Error: request for 'XXXX' is not yet fulfilled` without the fix.

### Risk

There is a risk that this adds a regression to the eleventy SSR plugin if something isn't quite right.

### CODE REVIEW

File: packages/labs/ssr/src/lib/module-loader.ts [hunk 1]
```diff
@@ -247,6 +247,19 @@ export class ModuleLoader {
...
     this.cache.set(modulePath, moduleRecord);
     const module = await modulePromise;
+    // Modules must be fully loaded before linking. Therefore `_loadModule` must
+    // also load its dependencies.
+    // Reference: https://tc39.es/ecma262/#table-abstract-methods-of-module-records
+    const {identifier} = module;
+    if (!/:\d+$/.test(identifier)) {
+      throw new Error('Unexpected file:// URL identifier without context ID');
+    }
+    const moduleReferrerPath = identifier.split(/:\d+$/)[0];
```
Comment (justinfagnani) [inline packages/labs/ssr/src/lib/module-loader.ts:257]: Definitely put the regex into a var if you're using it more than once.

File: packages/labs/ssr/src/lib/module-loader.ts [hunk 2]
```diff
@@ -247,6 +247,19 @@ export class ModuleLoader {
     };
     this.cache.set(modulePath, moduleRecord);
     const module = await modulePromise;
+    // Modules must be fully loaded before linking. Therefore `_loadModule` must
```
Comment (justinfagnani) [inline packages/labs/ssr/src/lib/module-loader.ts:250]: Let's add to the jsdoc comment for _loadModule that this load all dependencies too.

**Review by justinfagnani — APPROVED** [review]
Awesome work!

File: packages/labs/ssr/src/lib/module-loader.ts [hunk 3]
```diff
@@ -247,6 +247,19 @@ export class ModuleLoader {
...
+    // also load its dependencies.
+    // Reference: https://tc39.es/ecma262/#table-abstract-methods-of-module-records
+    const {identifier} = module;
+    if (!/:\d+$/.test(identifier)) {
```
[the remark this answers is not in the document]
Comment (AndrewJakubowicz) [reply packages/labs/ssr/src/lib/module-loader.ts]: I pulled out this repeated logic into a small util function. The context of that function should make it clearer what is happening.

File: packages/labs/ssr/src/lib/module-loader.ts [hunk 4]
```diff
@@ -247,6 +247,19 @@ export class ModuleLoader {
...
+      throw new Error('Unexpected file:// URL identifier without context ID');
```
[the remark this answers is not in the document]
Comment (AndrewJakubowicz) [reply packages/labs/ssr/src/lib/module-loader.ts]: I couldn't find any specific docs on what context ID is. But this method changes:

`/my/file/system/lit/lit-element.js:8` -> `/my/file/system/lit/lit-element.js`

Added more descriptive error message (with Internal Error prefix)

[the remark this answers is not in the document]
Comment (AndrewJakubowicz) [reply packages/labs/ssr/src/lib/module-loader.ts]: Ah, I figured it out. It's our own identifier scheme. We have the following method which creates these paths:

```
  private _getIdentifier(modulePath: string) {
    return `${modulePath}:${this._vmContextId}`;
  }
```

I've updated documentation to make this clearer in the future.

File: packages/labs/ssr/src/lib/module-loader.ts (hunk 1 above)
Comment (AndrewJakubowicz) [reply packages/labs/ssr/src/lib/module-loader.ts]: Done

**Review by augustjk — APPROVED** [review]
Nice!

**Pull request merged**